doc.go

```diff
@@ -18,121 +18,253 @@
-//   	})
+//   func SetupGateways(conn *sql.DB) (*UserGateway, *CommentGateway, *PostGateway, error) {
+//     // ...
 //   }
 //
+// The above is equivalent to,
+//
+//  type Gateways struct {
+//    dig.Out
+//
+//    Users    *UserGateway
+//    Comments *CommentGateway
+//    Posts    *PostGateway
+//  }
+//
+//  func SetupGateways(conn *sql.DB) (Gateways, error) {
+//    // ...
+//  }
+//
+// Optional Dependencies
+//
+// Constructors often don't have a hard dependency on some types and
+// are able to operate in a degraded state when that dependency is missing.
+// Dig supports declaring dependencies as optional by adding an
+// `optional:"true"` tag to fields of a dig.In struct.
+//
+// Fields in a dig.In structs that have the `optional:"true"` tag are treated
+// as optional by Dig.
 //
+//   type UserGatewayParams struct {
+//     dig.In
+//
+//     Conn  *sql.DB
+//     Cache *redis.Client `optional:"true"`
+//   }
+//
+// If an optional field is not available in the container, the constructor
+// will receive a zero value for the field.
+//
+//   func NewUserGateway(p UserGatewayParams, log *log.Logger) (*UserGateway, error) {
+//     if p.Cache != nil {
+//       log.Print("Logging disabled")
+//     }
+//     // ...
+//   }
+//
+// Constructors that declare dependencies as optional MUST handle the case of
+// those dependencies being absent.
+//
+// The optional tag also allows adding new dependencies without breaking
+// existing consumers of the constructor.
+//
+// Named Values
+//
+// Some use cases call for multiple values of the same type. Dig allows adding
+// multiple values of the same type to the container with the use of
+// `name:".."` tags on fields of dig.In and dig.Out structs.
+//
+// A constructor that produces a dig.Out struct can tag any field with
+// `name:".."` to have the corresponding value added to the graph under the
+// specified name.
+//
+//   type ConnectionResult {
+//     dig.Out
+//
+//     ReadWrite *sql.DB `name:"rw"`
+//     ReadOnly  *sql.DB `name:"ro"`
+//   }
+//
+//   func ConnectToDatabase(...) (ConnectionResult, error) {
+//     // ...
+//     return ConnectionResult{ReadWrite: rw, ReadOnly:  ro}, nil
+//   }
+//
+// Another constructor can consume these values by adding fields to a dig.In
+// struct with the same name AND type.
+//
+//   type GatewayParams struct {
+//     dig.In
+//
+//     WriteToConn  *sql.DB `name:"rw"`
+//     ReadFromConn *sql.DB `name:"ro"`
+//   }
+//
+// The name tag may be combined with the optional tag to declare the
+// dependency optional.
+//
+//   type GatewayParams struct {
+//     dig.In
+//
+//     WriteToConn  *sql.DB `name:"rw"`
+//     ReadFromConn *sql.DB `name:"ro" optional:"true"`
+//   }
+//
+//   func NewCommentGateway(p GatewayParams, log *log.Logger) (*CommentGateway, error) {
+//     if p.ReadFromConn == nil {
+//       log.Print("Warning: Using RW connection for reads")
+//       p.ReadFromConn = p.WriteToConn
+//     }
+//     // ...
+//   }
```

Comment:
Awesome - really appreciate the effort that went into this. Can we also file an issue to convert as many of these code samples as possible to testable examples?